### ISSUE TITLE
How does autotrain calculate the loss during the finetuning process?

### ISSUE THREAD

**Issue by xihajun** [huggingface/autotrain-advanced#345]
Thanks again for the nice tool! I'm exploring the AutoTrain library and am curious about the specifics of how the loss is calculated during the fine-tuning process. I noticed that the library seems to utilize the default trainer from the transformers library, as seen in the code snippet from the GitHub repository: https://github.com/huggingface/autotrain-advanced/blob/22944c0cbb1c5e0f0929407438e8a6ae34fe9ed2/src/autotrain/trainers/clm/__main__.py#L367-L375.

Additionally, it appears that AutoTrain sets certain default parameters for training, which are also outlined in the repository: https://github.com/huggingface/autotrain-advanced/blob/22944c0cbb1c5e0f0929407438e8a6ae34fe9ed2/src/autotrain/trainers/clm/__main__.py#L324-L348

https://github.com/huggingface/autotrain-advanced/blob/22944c0cbb1c5e0f0929407438e8a6ae34fe9ed2/src/autotrain/trainers/clm/params.py#L9-L54

I wonder if it is possible to know how does AutoTrain employ the default loss function provided by the transformers library, or does it adapt the loss function dynamically based on different model architectures? Please do correct me if I am looking at the wrong place.

Any clarification on this matter would be greatly appreciated! Thanks a lot for your time in advanced!

**Comment by shreydan** [comment]
@xihajun auto-regressive (causal LMs) models or Seq2Seq models have a `lm_head` layer which is used to predict which token comes next.
In Hugging Face models, during training when we provide the labels to the model, the loss is calculated (in the forward pass itself) using `nn.CrossEntropyLoss` for all language models. The inputs are the predictions from the `lm_head` layer and the targets are the provided labels.

Also, the labels are just inputs to the decoder shifted by 1 token, so that we know which token to predict after the current token.
```
inputs:                 100,24,43,53,24,84,0
right-shifted-labels:   24,43,53,24,84,0,0 # pad token = 0
```
during loss calculation, all pad tokens are replaced with -100 so the loss function ignores those tokens.

code examples: [gpt-2](https://github.com/huggingface/transformers/blob/8eb9e29d8dc8b8bd98b4dd48317d1d596ec548f3/src/transformers/models/gpt2/modeling_gpt2.py#L1099), [llama](https://github.com/huggingface/transformers/blob/8eb9e29d8dc8b8bd98b4dd48317d1d596ec548f3/src/transformers/models/llama/modeling_llama.py#L1059)

**Comment by xihajun** [comment]
@shreydan Thank you so much for the lovely explanation and pointing out to the code example! That's super helpful!

(note: for someone who might find this useful, `ignore_index=-100` is the default ignore index for PyTorch's CrossEntropyLoss: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
